### PR TITLE
8323576: [Windows] Fallthrough to ::abort instead of os::infinite_sleep for noreturn methods

### PR DESCRIPTION
**Pull request by TheShermanTanker**
os::infinite_sleep should only be used in very specific scenarios, and using it to satisfy compiler requirements for [[noreturn]] marked methods is a bad idea, since in the impossible case this will cause HotSpot to hang for a seemingly unknown reason. Fallthrough to ::abort instead, like other platforms do, to align Windows with them

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323576](https://bugs.openjdk.org/browse/JDK-8323576): [Windows] Fallthrough to ::abort instead of os::infinite_sleep for noreturn methods (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17366/head:pull/17366` \
`$ git checkout pull/17366`

Update a local copy of the PR: \
`$ git checkout pull/17366` \
`$ git pull https://git.openjdk.org/jdk.git pull/17366/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17366`

View PR using the GUI difftool: \
`$ git pr show -t 17366`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17366.diff">https://git.openjdk.org/jdk/pull/17366.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17366#issuecomment-1886620973)

### CODE REVIEW

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
We don't use `std::exit` or `std::abort`, we just use the C library methods.

Why not just use `ShouldNotReachHere()`?

Thanks

**Comment by TheShermanTanker** [comment]
I believe ShouldNotReachHere ultimately calls back into the code in question here, so doing that would ultimately result in a never ending recursion loop (I'm not at home right now however, so I cannot confirm whether this is true). This also aligns with the POSIX platforms, which all fall through to ::abort in their respective terminate methods

**Comment by dholmes-ora** [comment]
> I believe ShouldNotReachHere ultimately calls back into the code in question here

I don't think it does. This code underpins `os::exit` while `ShouldNotReachHere` should lead to an abort.

But okay to have windows call `::abort` like other platforms, just not `std::abort`.

**Comment by TheShermanTanker** [comment]
I think VMError::report\_and\_die calls os::exit and os::abort (both of which call exit\_process\_or\_thread), and vaguely also remember it calling raise\_fail\_fast, but to answer that question definitively, probably need Kim's confirmation. Can do on that second part, but out of curiosity, is there a difference between std:: and :: for C methods?

**Comment by dholmes-ora** [comment]
> is there a difference between std:: and :: for C methods?

There are differences between C++ exit and C exit, as the former performs additional C++ cleanup - that said I'm unclear if we actually end up executing the C or the C++ version of the function! But we don't use the `std::` namespace in this area. Also we have the `ALLOW_C_FUNCTION` macros to make it clear that it is correct to be calling something like `::exit` or `::abort` rather than `os::exit` or `os::abort`.

File: src/hotspot/os/windows/os_windows.cpp
Comment (TheShermanTanker) [reply src/hotspot/os/windows/os_windows.cpp:4238]: Ah, I was thinking the standard C method could be preferred over the Microsoft deprecated \_exit method, but I guess that's out of scope. I'm currently working on some build dev issue, I'll change this back later on

Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.cpp:4238]: Okay. The `_Exit` is the ISO C function, but POSIX also defines `_exit` (not that that is relevant to Windows). I can't see anything saying that Microsoft have deprecated `_exit` though:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/exit-exit-exit?view=msvc-170
If we were to change this I would still want the `ALLOW_C_FUNCTION`. Thanks

Comment (TheShermanTanker) [reply src/hotspot/os/windows/os_windows.cpp:4238]: I guess I can leave that \_Exit change for another day, it wouldn't make sense to have ALLOW\_C\_FUNCTION on something that is not banned by HotSpot after all

**Review by dholmes-ora — APPROVED** [review]
Looks good.

**Comment by TheShermanTanker** [comment]
Thanks David. Is Kim around to verify whether ShouldNotReachHere and friends call recursively into os::exit or os::abort?

**Comment by TheShermanTanker** [comment]
Don't waiting for @kimbarrett's approval

**Comment by TheShermanTanker** [comment]
@dholmes-ora Sorry for the page, is Kim around?

**Comment by dholmes-ora** [comment]
@TheShermanTanker Kim may have GH notifications disabled, as I do, so may not have seen your earlier page. Trying paging via JBS or else direct email. AFAIK Kim is around.

**Comment by TheShermanTanker** [comment]
Ah, right. Thanks for the tip!

File: src/hotspot/os/windows/os_windows.cpp
Comment (kimbarrett) [reply src/hotspot/os/windows/os_windows.cpp]: For the record, `::_exit` _is_ forbidden.
https://github.com/openjdk/jdk/blob/7ac2f914aadf7fb12b7258e54efae959e15d6721/src/hotspot/share/utilities/globalDefinitions.hpp#L191
It might be deprecated as part of the optional defining and deprecation of POSIX names:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/compatibility?view=msvc-170

**Comment by kimbarrett** [comment]
> Thanks David. Is Kim around to verify whether ShouldNotReachHere and friends call recursively into os::exit or os::abort?

I expect it does.  It just calls report_vm_error like lots of other things, which calls VMError::report_and_die, and so on.

**Review by kimbarrett — APPROVED** [review]
Looks good.

I wondered whether we should be concerned about the Microsoft-specific
behavior of abort.
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/abort?view=msvc-170

But these are supposed to be unreachable code paths.  And we already use abort
in other places.

**Comment by TheShermanTanker** [comment]
Thanks David and Kim for the reviews!

**Pull request closed without merging**
